### PR TITLE
Enable state integrity check

### PR DESCRIPTION
**Pull request by gitmp01**

### CODE REVIEW

File: android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java
Comment (gitmp01) [inline android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java:64]: We don't need this one anymore as the extension is loaded directly in the SQLite library now

File: setup-plugdev-perm.sh
Comment (gitmp01) [inline setup-plugdev-perm.sh:1]: This file is useful when you connect the device to your laptop in order to set the correct udev permission withing a blink of an eye (needs `sudo` permissions)

Comment (gitmp01) [reply setup-plugdev-perm.sh:1]: I don't know if it'll still work with the samsung though, let me know in case, probably there would be the need to replace `google` with `samsung` or something like that

File: android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java
Comment (gskapka) [inline android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java]: Do we still need this logging?

File: android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java
Comment (gskapka) [inline android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java]: We can't actually reach here without this already being false, so this is redundant. (See line #234)

File: android/app/src/main/java/com/ptokenssentinelandroidapp/strongbox/Strongbox.java
Comment (gskapka) [inline android/app/src/main/java/com/ptokenssentinelandroidapp/strongbox/Strongbox.java:348]: Ah, so these were public for the state hash signing in the `DatabaseWiring` class.

I made them private when I had to expose an interface to this module to allow rust to be able to call stuff in it successfully.

File: setup-plugdev-perm.sh
Comment (gskapka) [inline setup-plugdev-perm.sh]: I know you commented in this PR what this is doing, but for later when that's long forgotten, it'd be useful to have that documented in here too.

File: android/sqlite3/README.md
Comment (gskapka) [inline android/sqlite3/README.md:1]: _Great_ update to this whole document!! Good idea doing it now too when you're fresh from actually debugging this missing-library issue 😅

**Review by gskapka — DISMISSED** [review]
Approved as is, but with a couple of minor comments in line that you can change at your discretion.

File: android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java
Comment (gitmp01) [reply android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java]: Nope, I'll remove it

File: android/app/src/main/java/com/ptokenssentinelandroidapp/strongbox/Strongbox.java
Comment (gitmp01) [reply android/app/src/main/java/com/ptokenssentinelandroidapp/strongbox/Strongbox.java:348]: Ah I see, do you think they could break anything? They shouldn't in my opinion

File: android/sqlite3/README.md
Comment (gitmp01) [reply android/sqlite3/README.md:1]: Yeah, I'm not going to make the same mistake twice :sweat_smile:

File: android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java
Comment (gitmp01) [reply android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java]: good catch

**Comment by gskapka** [comment]
Dismissing the review because I'm getting app crashes when testing now. 

It appears to be the first call to `startTransaction` failing due to verification of the state hash failing (presumably because it doesn't exist yet). 

Investigating more, but here are pertinent logs for now:

```
11-10 13:47:10.073 24163 24253 I sentinel_strongbox::a..: handling web socket message...
11-10 13:47:10.073 24163 24253 I com.ptokenssentinelandroidapp.database.DatabaseWiring: starting db tx...
11-10 13:47:10.075 24163 24253 W com.ptokenssentinelandroidapp.database.Operations: failed to get the signature from the file, ok if first run!
11-10 13:47:10.202 24163 24253 I com.ptokenssentinelandroidapp.database.DatabaseWiring: current state hashPCWZiz8loRyizz3FsbbXj54mJ/M=
11-10 13:47:10.203 24163 24253 E com.ptokenssentinelandroidapp.database.DatabaseWiring: signed state hash verification failed!
11-10 13:47:10.203 24163 24253 E com.ptokenssentinelandroidapp.database.DatabaseWiring: com.ptokenssentinelandroidapp.strongbox.StrongboxException: ✘ Missing signature for existing state!
```

I see the log that says this is okay if it's the first run, but I'm not seeing handling for this case (which there should be imo).

Seems to be hitting the first arm of this:

```java
        if (!signatureExists && hashExists) {
            throw new StrongboxException("✘ Missing signature for existing state!");
        } else if (signatureExists && !hashExists) {
            throw new StrongboxException("✘ Existing signature for missing state!");
        } else if (!signatureExists) {
            Log.i(TAG, "first run!");
            return;
        }

```

**Comment by gskapka** [comment]
Can also confirm the same behaviour & same error trace after a fresh re-install of the app on the device.

**Comment by gskapka** [comment]
Extra notes. This is testing on the pixel 3a so could there perchance be a valid state hash saved from previous v2 development? Meaning that then this case:

```java
!signatureExists && hashExists
```

which _should_ normally be entirely impossible, could in fact be possible?

Though if it were a vestige from previous apps, I'd presume an old signature to be present too.

**Comment by gitmp01** [comment]
No it can't since the file is saved into the app path, so it's different for each app

**Comment by gskapka** [comment]
Okay, that's that ruled out then.

This all also leads to a bigger question: What do we plan to do should the app reach some sort of state whereby the db integrity cannot be verified? How to portray that to the user? And what do suggest as a mitigation?

**Comment by gskapka** [comment]
Okay so the above logs I've posted are the start of the cause of the issue, but they're not why the app crashes. 

The reason the app crashes is because the `startTransaction` function fails (due to db integrity verification failing), returning an error to the rust core, which handles said error, which handling results in a call from the core to `cancelTransaction` (as it does for any errors it encounters, since we always want to roll back the db state in these cases).

That rust-caused `cancelTransaction` fxn call then fails because of the `START_DB_TX_IN_PROGRESS` is left `true` when the integrity checks failed during the `startTransaction` fxn.

And these sort of errors are after we're back out of the rust handling, and thus the java exceptions have no handling and thus the app crashes. 

So I've handled that flag in the `catch` arm of the verification call in `startTransaction` java fxn, and now have a new error:

```java
11-10 14:51:51.029 26371 26426 F tinelandroidap: java_vm_ext.cc:579] JNI DETECTED ERROR IN APPLICATION: JNI NewStringUTF called with pending exception java.lang.IllegalStateException: Cannot perform this operation because there is no current transaction.

```

To mitigate that, I've added in a check `db.inTransaction()` inside the `cancelTransaction` fxn.

Sure enough, we no longer get app crashes when db integrity checks fail for whatever reason.

However we're now in the state where we just get a generic error returned to the core:

```rust
      "SentinelError": "jni error: Java exception was thrown"
```

which is from the exception that occurred during `startTransaction` in rust. This leaves us with two remaining questions:

1) The one I posted about about how to handle the general case of db integrity not being able to be verified for some reason and 
2) How to recover from such a situation?

(Note that I've pushed the above fixes to this branch)

**Comment by gitmp01** [comment]
Can be a multithreaded issue? Cen you check it's working with just the init calls and syncers disabled?

**Comment by gskapka** [comment]
It's nothing to do with threading - I was calling a single fxn to the core with everything else disabled in order to get the errors shown above.

The android core is _not_ multithreaded, and the websocket pathway to it enforces that single queue from the multi-threaded outside sentinel app.

I've pushed two more commits to augment the dev `dropDb()` function, which now also deletes any existing db-state-hashes when dropping the db.

Sure enough, dropping the db then re-installing and re-initializing a sentinel has it up and running smoothly again, syncing nicely, with db state integrity checks passing for each and every db transaction. 

So I think the open questions above are outside of the purview of this PR, and this PR is complete w/r/t re-implementing db state integrity checks.

**Comment by gskapka** [comment]
So I'll await your approval on the commits I've pushed to this PR, and once approved we can get it merged into the main sentinel dev branch if that's okay with you?

**Comment by gitmp01** [comment]
Aah sorry, it wasn't clear to me. Perfect, I can look at your commits in half an hour

I agree on the above obvs

**Comment by gskapka** [comment]
Note also that whilst I've not benched anything (don't have the means currently), I'm not seeing/noticing any degraded performance due to these new overheads during my testing, so that's good. (Albeit v subjective 😅 )

File: android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java
Comment (gitmp01) [inline android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java:369]: :+1:

File: android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java
Comment (gitmp01) [inline android/app/src/main/java/com/ptokenssentinelandroidapp/database/DatabaseWiring.java:149]: Good catch!

**Review by gitmp01 — COMMENTED** [review]
New commits are good to me!

File: android/app/src/main/java/com/ptokenssentinelandroidapp/strongbox/Strongbox.java
Comment (gskapka) [reply android/app/src/main/java/com/ptokenssentinelandroidapp/strongbox/Strongbox.java:348]: Nope - was just explaining why I changed them in the first place.

**Pull request merged**
